### ISSUE TITLE
[Feature]: Faster apply_top_k_top_p without scatter

### ISSUE THREAD

**Issue by xyz2606** [vllm-project/vllm#27710]
### 🚀 The feature, motivation and pitch

In `apply_top_k_top_p`, the following code unpermutes the sorted logits:
`logits = logits_sort.scatter(dim=-1, index=logits_idx, src=logits_sort)`.
Then we randomly sample a token from the distribution by
`return probs.div(q).argmax(dim=-1).view(-1)`.

The two steps above are equivalent to 
`idx_permuted = probs.div(q).argmax(dim=-1).view(-1)`
`return logits_idx[idx_permuted]` which (almostly) saves a `scatter` operation. `idx_permuted` is much smaller than `logits_sort`.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
